Slim/Plugin/ExtendedBrowseModes/Settings.pm

```diff
@@ -55,8 +56,14 @@ sub handler {
 	if ($params->{'saveSettings'} && !$class->needsClient) {
 		# custom role handling
 		my $currentRoles = $serverPrefs->get('userDefinedRoles');
+
+		my $id = 20;
```

Comment:
```perl
my $id = max(map { $_->{id} } values %$currentRoles) + 1;
```
?

Besides that "simplification" (untested!) I believe this kind of code belongs in `S::Schema::Contributor`. I actually believe nothing should read/write `preferences('server')->get('userDefinedRoles')` outside that module. We should have accessors there to do the manipulation. Otherwise this kind of logic is spread across too many places IMHO.

Reply:
I'll test the simplification. I do need to become more confident in using `map`.

Otherwise, I'm not quite sure what you mean: manipulating the user-defined roles makes no sense outside of the handler for the prefs page. Am I missing something (very possible!)?

Reply:
That's correct. But there's similar code when reading the pref in various places. Just putting it in one place seems cleaner to me.

Reply:
So you want me to encapsulate all of lines 56 to 97 in Slim::Schema::Contributor?  If so, it's a lot of parameter passing.

If not, what exactly?

Reply:
Maybe I'm oversimplifying my assumptions. What parameters would you need? Name and "include" flag?

No need to evaluate the `$id` here. Just tell the schema to add a role. It would do the check locally.

Reply:
This reminded me we need to ensure the user-defined roles are created starting with id 21, so I've had to amend your above simplification above so it works when creating the first user-defined role. (The test >20 is used in LMS and Material already, to distinguish user-defined roles.)

At the moment, we're rebuilding the user-defined roles pref map as a whole, then replacing wholesale, before triggering Slim::Schema::Contributor->initializeRoles() to rebuild the contributor-role maps/arrays from the newly-written pref.

I don't think we should move the actual pref setting outside of the settings handler.

Not sure where this leaves us?